### PR TITLE
[SPARK-31189][SQL][DOCS] Fix errors and missing parts for datetime pattern document

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Fix errors and missing parts for datetime pattern document
1. The pattern we use is similar to DateTimeFormatter and SimpleDateFormat but not identical. So we shouldn't use any of them in the API docs but use a link to the doc of our own.
2. Some pattern letters are missing
3. Some pattern letters are explicitly banned - Set('A', 'c', 'e', 'n', 'N')
4. the second fraction pattern different logic for parsing and formatting

### Why are the changes needed?

fix and improve doc
### Does this PR introduce any user-facing change?

yes, new and updated doc
### How was this patch tested?

pass Jenkins
viewed locally with `jekyll serve`
image

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @xuanyuanking thanks

**Comment by maropu** [comment]
also cc: @MaxGekk

File: docs/sql-ref-datetime-pattern.md [hunk 1]
```diff
@@ -21,11 +21,26 @@ license: |
...
 There are several common scenarios for datetime usage in Spark:
 
-- CSV/JSON datasources use the pattern string for parsing and formatting time content.
+- CSV/JSON datasources use the pattern string for parsing and formatting datetime content.
 
-- Datetime functions related to convert string to/from `DateType` or `TimestampType`. For example, unix_timestamp, date_format, to_unix_timestamp, from_unixtime, to_date, to_timestamp, from_utc_timestamp, to_utc_timestamp, etc.
+- Datetime functions related to convert `StringType` to/from `DateType` or `TimestampType`.
+  For example, `unix_timestamp`, `date_format`, `to_unix_timestamp`, `from_unixtime`, `to_date`, `to_timestamp`, `from_utc_timestamp`, `to_utc_timestamp`, etc.
+
+Spark uses [java.time.format.DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: I wouldn't mention these 2 java classes at all. Users don't care how we implement it, but what the behavior is.

File: docs/sql-ref-datetime-pattern.md [hunk 2]
```diff
@@ -147,6 +168,30 @@ Spark uses the below letters in date and timestamp parsing and formatting:
...
   <td> 978 </td>
 </tr>
+<tr>
+  <td> <b>A</b> </td>
+  <td> milli-of-day </td>
+  <td> number </td>
+  <td> 1234 </td>
+</tr>
+<tr>
+  <td> <b>n</b> </td>
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: simple n/N?

Comment (cloud-fan) [reply docs/sql-ref-datetime-pattern.md]: BTW does spark 2.4 support it?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala [hunk 3]
```diff
@@ -253,11 +253,13 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
...
    * created by `PERMISSIVE` mode. This overrides `spark.sql.columnNameOfCorruptRecord`.</li>
    * <li>`dateFormat` (default `yyyy-MM-dd`): sets the string that indicates a date format.
-   * Custom date formats follow the formats at `java.time.format.DateTimeFormatter`.
+   * Custom date formats follow the formats at
+   * http://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html.
    * This applies to date type.</li>
    * <li>`timestampFormat` (default `yyyy-MM-dd'T'HH:mm:ss.SSSXXX`): sets the string that
    * indicates a timestamp format. Custom date formats follow the formats at
-   * `java.time.format.DateTimeFormatter`. This applies to timestamp type.</li>
+   * http://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html.
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala]: Shall we use `<a href="http://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html"> Datetime patterns</a>`?

File: docs/sql-ref-datetime-pattern.md (hunk 1 above)
Comment (yaooqinn) [reply docs/sql-ref-datetime-pattern.md]: removed

File: docs/sql-ref-datetime-pattern.md [hunk 4]
```diff
@@ -199,11 +231,15 @@ The count of pattern letters determines the format.
 
 - Number/Text: If the count of pattern letters is 3 or greater, use the Text rules above. Otherwise use the Number rules above.
 
-- Fraction: Outputs the micro-of-second field as a fraction-of-second. The micro-of-second value has six digits, thus the count of pattern letters is from 1 to 6. If it is less than 6, then the micro-of-second value is truncated, with only the most significant digits being output.
+- Fraction: Use `'S..S'` pattern for parsing and formatting fraction of second.
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: `S..S` is a bit confusing. How about `Use one or more contiguous 'S' characters ...`

File: docs/sql-ref-datetime-pattern.md [hunk 5]
```diff
@@ -199,11 +231,15 @@ The count of pattern letters determines the format.
...
+  For parsing, the acceptable fraction length can be [1, `'S..S'`.length(<=9)] and with `[]` surrounded can be [0, `'S..S'`.length(<=9)].
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: `[1, number of contiguous 'S']`

Comment (cloud-fan) [reply docs/sql-ref-datetime-pattern.md]: And we don't need to mention `[]`, as it's semantic is well defined and orthogonal to the second fraction.

File: docs/sql-ref-datetime-pattern.md [hunk 6]
```diff
@@ -199,11 +231,15 @@ The count of pattern letters determines the format.
...
+  For formatting, with or without `[]` surrounded, the fraction length would be padded to `'S..S'`.length(<=9) with zeros.
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: ditto

File: docs/sql-ref-datetime-pattern.md [hunk 7]
```diff
@@ -189,6 +209,18 @@ Spark uses the below letters in date and timestamp parsing and formatting:
   <td> literal </td>
   <td> ' </td>
 </tr>
+<tr>
+  <td> <b>[</b> </td>
+  <td> optional section start </td>
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: shall we further expalin it in the detail section below?

Comment (yaooqinn) [reply docs/sql-ref-datetime-pattern.md]: explanations added

File: docs/sql-ref-datetime-pattern.md [hunk 8]
```diff
@@ -199,18 +231,27 @@ The count of pattern letters determines the format.
 
 - Number/Text: If the count of pattern letters is 3 or greater, use the Text rules above. Otherwise use the Number rules above.
 
-- Fraction: Outputs the micro-of-second field as a fraction-of-second. The micro-of-second value has six digits, thus the count of pattern letters is from 1 to 6. If it is less than 6, then the micro-of-second value is truncated, with only the most significant digits being output.
+- Fraction: Use one or more contiguous `'S'` characters, e,g `SSSSSS`, to parse and format fraction of second.
+  For parsing, the acceptable fraction length can be [1, the number of contiguous 'S'].
+  For formatting, the fraction length would be padded to the number of contiguous 'S' with zeros.
+  Spark supports datetime with max precision to micro-of-second which has six significant digits at most, but can parse nano-of-second field with exceed part truncated.
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: `but can parse nano-of-second` does it indicates that we accept at most 9 contiguous "S"?

File: docs/sql-ref-datetime-pattern.md [hunk 9]
```diff
@@ -199,18 +231,27 @@ The count of pattern letters determines the format.
...
 
 - Offset Z: This formats the offset based on the number of pattern letters. One, two or three letters outputs the hour and minute, without a colon, such as '+0130'. The output will be '+0000' when the offset is zero. Four letters outputs the full form of localized offset, equivalent to four letters of Offset-O. The output will be the corresponding localized offset text if the offset is zero. Five letters outputs the hour, minute, with optional second if non-zero, with colon. It outputs 'Z' if the offset is zero. Six or more letters will fail.
 
+- Optional section start and end: Use `[]` to define an optional section and maybe nested.All elements in the optional section are treated as optional.
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: nit: a space between dot and `All`

Comment (cloud-fan) [reply docs/sql-ref-datetime-pattern.md]: `All elements in the optional section are treated as optional.` Actually this whole sentence seems not informative...

File: docs/sql-ref-datetime-pattern.md [hunk 10]
```diff
@@ -199,18 +231,27 @@ The count of pattern letters determines the format.
...
+  During formatting, the section will be output when data is available.
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: It's hard to understand `when data is available`. For `ss[.SSS]` we always output the second fraction even if it's 0, right?

File: docs/sql-ref-datetime-pattern.md [hunk 11]
```diff
@@ -199,18 +231,27 @@ The count of pattern letters determines the format.
...
+  During parsing, the whole section may be missing from the parsed string.
+  An optional section is started by `[` and ended using `]` (or at the end of the pattern).
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: So `abc[ef` is valid? interesting...

File: python/pyspark/sql/functions.py [hunk 12]
```diff
@@ -973,8 +973,9 @@ def date_format(date, format):
     format given by the second argument.
 
     A pattern could be for instance `dd.MM.yyyy` and could return a string like '18.03.1993'. All
-    pattern letters of the Java class `java.time.format.DateTimeFormatter` can be used.
+    pattern letters of `datetime pattern`_. can be used. # noqa
```
Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: what does `# noqa` means?

File: python/pyspark/sql/functions.py [hunk 13]
```diff
@@ -973,8 +973,9 @@ def date_format(date, format):
...
 
+    .. _datetime pattern: /sql-ref-datetime-pattern.html
```
Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: cc @HyukjinKwon do you know what will show up in the python doc?

File: python/pyspark/sql/functions.py [hunk 14]
```diff
@@ -1191,8 +1192,7 @@ def months_between(date1, date2, roundOff=True):
 @since(2.2)
 def to_date(col, format=None):
     """Converts a :class:`Column` into :class:`pyspark.sql.types.DateType`
-    using the optionally specified format. Specify formats according to
-    `DateTimeFormatter <https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html>`_. # noqa
+    using the optionally specified format. Specify formats according to `datetime pattern`_. # noqa
```
Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: does `datetime pattern`_ work here? The URL is not specified in this method.

File: python/pyspark/sql/functions.py [hunk 15]
```diff
@@ -1215,8 +1215,7 @@ def to_date(col, format=None):
 @since(2.2)
 def to_timestamp(col, format=None):
     """Converts a :class:`Column` into :class:`pyspark.sql.types.TimestampType`
-    using the optionally specified format. Specify formats according to
-    `DateTimeFormatter <https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html>`_. # noqa
+    using the optionally specified format. Specify formats according to `datetime pattern`_. # noqa
```
Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: ditto

File: python/pyspark/sql/functions.py (hunk 14 above)
Comment (yaooqinn) [reply python/pyspark/sql/functions.py]: I checked the `partition discovery` here http://spark.apache.org/docs/3.0.0-preview2/api/python/pyspark.sql.html#module-pyspark.sql.streaming. Some take effect but others don't. BUT ALSO /xxx.html seems not to point to the right page.

File: docs/sql-ref-datetime-pattern.md (hunk 11 above)
Comment (yaooqinn) [reply docs/sql-ref-datetime-pattern.md]: yea, it is documented in java doc

File: docs/sql-ref-datetime-pattern.md (hunk 10 above)
Comment (yaooqinn) [reply docs/sql-ref-datetime-pattern.md]: yes

File: python/pyspark/sql/functions.py (hunk 14 above)
Comment (yaooqinn) [reply python/pyspark/sql/functions.py]: The full url is rewritten to http://spark.apache.org/sql-data-sources-parquet.html#partition-discovery

File: docs/sql-ref-datetime-pattern.md (hunk 10 above)
Comment (yaooqinn) [reply docs/sql-ref-datetime-pattern.md]: I guess java doc means that the datetime string you parse may be invalid and will not output. so maybe we can change available to valid

Comment (cloud-fan) [reply docs/sql-ref-datetime-pattern.md]: SGTM

File: docs/sql-ref-datetime-pattern.md [hunk 16]
```diff
@@ -199,18 +231,28 @@ The count of pattern letters determines the format.
 
 - Number/Text: If the count of pattern letters is 3 or greater, use the Text rules above. Otherwise use the Number rules above.
 
-- Fraction: Outputs the micro-of-second field as a fraction-of-second. The micro-of-second value has six digits, thus the count of pattern letters is from 1 to 6. If it is less than 6, then the micro-of-second value is truncated, with only the most significant digits being output.
+- Fraction: Use one or more contiguous `'S'` characters, e,g `SSSSSS`, to parse and format fraction of second.
+  For parsing, the acceptable fraction length can be [1, the number of contiguous 'S'].
+  For formatting, the fraction length would be padded to the number of contiguous 'S' with zeros.
+  Spark supports datetime with max precision to micro-of-second which has 6 significant digits at most, but can parse nano-of-second field which can accept 9 contiguous 'S' at most with exceeded part truncated.
```
Comment (cloud-fan) [inline docs/sql-ref-datetime-pattern.md]: Let's simplify it a bit
```
Spark supports datetime of micro-of-second precision, which has up to 6 significant digits. but can parse
nano-of-second with exceeded part truncated.
```
Let's mention the length limit when we introduce 'S', as it applies to formatting as well
```
Fraction: Use one or more contiguous `'S'` characters (up to 9 contiguous 'S'), e,g `SSSSSS`, ...
```

File: python/pyspark/sql/functions.py [hunk 17]
```diff
@@ -973,8 +973,9 @@ def date_format(date, format):
     format given by the second argument.
 
     A pattern could be for instance `dd.MM.yyyy` and could return a string like '18.03.1993'. All
-    pattern letters of the Java class `java.time.format.DateTimeFormatter` can be used.
+    pattern letters of `datetime pattern`_. can be used.
 
+    .. _datetime pattern: http://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html
```
Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: https?

File: docs/sql-ref-datetime-pattern.md (hunk 16 above)
Comment (yaooqinn) [reply docs/sql-ref-datetime-pattern.md]: Maybe more simpler
```
 Use one or more  (up to 9) contiguous `'S'` characters, e,g `SSSSSS`, ...
```

File: docs/sql-ref-datetime-pattern.md [hunk 18]
```diff
@@ -177,6 +191,12 @@ Spark uses the below letters in date and timestamp parsing and formatting:
   <td> offset-Z </td>
   <td> +0000; -0800; -08:00; </td>
 </tr>
+<tr>
+  <td> <b>p</b> </td>
+  <td> pad next field with spaces </td>
```
Comment (xuanyuanking) [inline docs/sql-ref-datetime-pattern.md]: Should we also add `Pad modifier` in the detail section if we decided to announce the support for `p`?

File: python/pyspark/sql/streaming.py [hunk 19]
```diff
@@ -491,6 +490,7 @@ def json(self, path, schema=None, primitivesAsString=None, prefersDecimal=None,
                                     disables `partition discovery`_.
 
         .. _partition discovery: /sql-data-sources-parquet.html#partition-discovery
```
Comment (cloud-fan) [inline python/pyspark/sql/streaming.py]: So this is wrong right? @HyukjinKwon can you check the python doc and see if the link is corrected?

File: docs/sql-ref-datetime-pattern.md (hunk 18 above)
Comment (cloud-fan) [reply docs/sql-ref-datetime-pattern.md]: +1, or we can forbid it if it's hard to document. I never used `p` before...

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

File: python/pyspark/sql/streaming.py (hunk 19 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/streaming.py]: This is fine - it's a way of reusing a link

Comment (HyukjinKwon) [reply python/pyspark/sql/streaming.py]: Sorry for my late response.

Comment (yaooqinn) [reply python/pyspark/sql/streaming.py]: the links are malformed, please check  http://spark.apache.org/docs/3.0.0-preview2/api/python/pyspark.sql.html#module-pyspark.sql.streaming with `partition discovery` content

Comment (HyukjinKwon) [reply python/pyspark/sql/streaming.py]: Oh, right. The root directory is different when it's released. Sure, seems like it's best to stick to the absolute links then. The syntax itself is fine as a relative link

Comment (HyukjinKwon) [reply python/pyspark/sql/streaming.py]: @yaooqinn can you make a PR to fix these?

Comment (yaooqinn) [reply python/pyspark/sql/streaming.py]: OK, I'll check it

**Pull request closed without merging**